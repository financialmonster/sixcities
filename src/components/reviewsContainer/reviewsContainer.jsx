import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {bindActionCreators} from 'redux';

import {fetchReviews} from '../../actions/reviews.js';
import Spinner from '../spinner/spinner.jsx';
import FormContainer from '../formContainer/formContainer.jsx';
import Reviews from '../reviews/reviews.jsx';
import ErrorIndicator from '../errorIndicator/errorIndicator.jsx';

const ReviewsContainer = ({reviews, isReviewsLoading, isReviewsError, id, fetchReviews}) => {
    useEffect(() => {
        fetchReviews(id);
    }, [fetchReviews, id]);
    
    if(isReviewsLoading) {
        return <Spinner />;
    } 
    if(isReviewsError) {
        return <ErrorIndicator operation="loading of the list of reviews" />;
    }

    return (
        <section className="property__reviews reviews">
            <h2 className="reviews__title">
                Reviews &middot; <span className="reviews__amount">{reviews.length}</span>
            </h2>
            <ul className="reviews__list">
                <Reviews reviews={reviews} />

                <FormContainer id={id} />
            </ul>
        </section>
    );
}

ReviewsContainer.propTypes = {
    id: PropTypes.number.isRequired,
    fetchReviews: PropTypes.func.isRequired,
    isReviewsError: PropTypes.bool.isRequired,
    isReviewsLoading: PropTypes.bool.isRequired,
    reviews: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        user: PropTypes.shape({
            id: PropTypes.number.isRequired,
            is_pro: PropTypes.bool.isRequired,
            name: PropTypes.string.isRequired,
            avatar_url: PropTypes.string.isRequired
        }),
        rating: PropTypes.number.isRequired,
        comment: PropTypes.string.isRequired,
        date: PropTypes.string.isRequired
    }))
}

const mapStateToProps = ({reviews: {isReviewsError, isReviewsLoading, reviews}}) => {
    return {
        isReviewsError,
        isReviewsLoading,
        reviews
    }
}

const mapDispatchToProps= (dispatch) => {
    return {
        fetchReviews: bindActionCreators(fetchReviews, dispatch)
    }
}

export {ReviewsContainer};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewsContainer);
