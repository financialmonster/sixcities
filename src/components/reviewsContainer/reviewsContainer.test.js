import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import {ReviewsContainer} from './reviewsContainer.jsx';

const reviews = [{id: 1, rating: 3, date: `2020-04-13T20:25:41.450Z`, user: {id: 15, is_pro: false, name: `Kendall`, 
    avatar_url: `https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/6.jpg`},
    comment: `I stayed here for one night and it was an unpleasant experience.`},
    {id: 2, rating: 2, date: `2020-04-13T20:25:41.450Z`, user: {id: 13, is_pro: false, name: `Zak`,
    avatar_url: `https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/4.jpg`},
    comment: `What an amazing view! The house is stunning and in an amazing location. The large glass wall had an
    amazing view of the river!`}
];

describe(`Correctly renders the ReviewsContainer component`, () => {
    it(`when isReviewsLoading property is equal true and isReviewsError property is equal false`, () => {
        const tree = renderer.create(
            <ReviewsContainer reviews={reviews} isReviewsLoading={true} isReviewsError={false} id={5}
                fetchReviews={() => {}}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isReviewsError property is equal true and isReviewsLoading property is equal false`, () => {
        const tree = renderer.create(
            <ReviewsContainer reviews={reviews} isReviewsLoading={false} isReviewsError={true} id={5}
                fetchReviews={() => {}}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when both isReviewsError and isReviewsLoading properties are equal false`, () => {
        const mockStore = configureStore([]);
        const store = mockStore({
            auth: {isLoggedIn: false, isAuthStatusLoading: false , isAuthStatusError: false},
            reviews: {isReviewsError: false, isReviewsLoading: false, isReviewError: false, isReviewLoading: false,
                reviews: []
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <ReviewsContainer reviews={reviews} isReviewsLoading={false} isReviewsError={false} id={5}
                    fetchReviews={() => {}}
                />
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
