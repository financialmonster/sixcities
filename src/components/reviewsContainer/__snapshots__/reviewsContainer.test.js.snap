// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders the ReviewsContainer component when both isReviewsError and isReviewsLoading properties are equal false 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/6.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Kendall
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          I stayed here for one night and it was an unpleasant experience.
        </p>
        <time
          className="reviews__time"
          dateTime="2020-4-14"
        >
          April 2020
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://htmlacademy-react-3.appspot.com/six-cities/static/avatar/4.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Zak
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "40%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          What an amazing view! The house is stunning and in an amazing location. The large glass wall had an
    amazing view of the river!
        </p>
        <time
          className="reviews__time"
          dateTime="2020-4-14"
        >
          April 2020
        </time>
      </div>
    </li>
  </ul>
</section>
`;

exports[`Correctly renders the ReviewsContainer component when isReviewsError property is equal true and isReviewsLoading property is equal false 1`] = `
<div
  className="cities__status-wrapper tabs__content"
>
  <b
    className="cities__status"
  >
    We're sorry
  </b>
  <p
    className="cities__status-description"
  >
    It's been an error during the process of 
    loading of the list of reviews
    . Please, reload the page and try again
  </p>
</div>
`;

exports[`Correctly renders the ReviewsContainer component when isReviewsLoading property is equal true and isReviewsError property is equal false 1`] = `
<div
  className="spinner"
>
  <div
    className="spinnerContent"
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;
